File: plc_register.js
#!/usr/bin/env bun

import { secp256k1 } from '@noble/curves/secp256k1';
import { p256 } from '@noble/curves/p256';
import { sha256 } from '@noble/hashes/sha256';
import { encode as dagCborEncode } from '@ipld/dag-cbor';
import bs58 from 'bs58';
import { Command } from 'commander';
import { mkdirSync, writeFileSync } from 'node:fs';
import { resolve } from 'node:path';

const MC_P256_PUB = new Uint8Array([0x80, 0x24]);
const MC_K256_PUB = new Uint8Array([0xe7, 0x01]);
const N_K256 = 0xfffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364141n;
const N_P256 = 0xffffffff00000000ffffffffffffffffbce6faada7179e84f3b9cac2fc632551n;

const K256_ALG_ID = new Uint8Array([
  0x30, 0x10, 0x06, 0x07, 0x2a, 0x86, 0x48, 0xce, 0x3d, 0x02, 0x01, 0x06, 0x05, 0x2b, 0x81,
  0x04, 0x00, 0x0a,
]);

const P256_ALG_ID = new Uint8Array([
  0x30, 0x13, 0x06, 0x07, 0x2a, 0x86, 0x48, 0xce, 0x3d, 0x02, 0x01, 0x06, 0x08, 0x2a, 0x86,
  0x48, 0xce, 0x3d, 0x03, 0x01, 0x07,
]);

function ensureDir(path) {
  mkdirSync(path, { recursive: true });
}

function b64urlNopad(data) {
  return Buffer.from(data).toString('base64url').replace(/=+$/g, '');
}

function base32LowerNopad(data) {
  const alphabet = 'abcdefghijklmnopqrstuvwxyz234567';
  let bits = 0;
  let value = 0;
  let result = '';
  for (const byte of data) {
    value = (value << 8) | byte;
    bits += 8;
    while (bits >= 5) {
      result += alphabet[(value >>> (bits - 5)) & 31];
      bits -= 5;
    }
  }
  if (bits > 0) {
    result += alphabet[(value << (5 - bits)) & 31];
  }
  return result;
}

function getCurve(curve) {
  if (curve === 'k256') {
    return { ec: secp256k1, order: N_K256, mcPrefix: MC_K256_PUB, algId: K256_ALG_ID };
  }
  if (curve === 'p256') {
    return { ec: p256, order: N_P256, mcPrefix: MC_P256_PUB, algId: P256_ALG_ID };
  }
  throw new Error("curve must be 'k256' or 'p256'");
}

function didKeyForPub(curve, compressedPubkey) {
  const { mcPrefix } = getCurve(curve);
  const prefixed = new Uint8Array(mcPrefix.length + compressedPubkey.length);
  prefixed.set(mcPrefix);
  prefixed.set(compressedPubkey, mcPrefix.length);
  return 'did:key:z' + bs58.encode(prefixed);
}

function lowS(curve, r, s) {
  const { order } = getCurve(curve);
  if (s > order / 2n) {
    s = order - s;
  }
  return [r, s];
}

function bigintToBytes(n, length) {
  const hex = n.toString(16).padStart(length * 2, '0');
  if (hex.length > length * 2) {
    throw new Error(`integer too large for ${length} bytes`);
  }
  const bytes = new Uint8Array(length);
  for (let i = 0; i < length; i++) {
    bytes[i] = parseInt(hex.slice(i * 2, i * 2 + 2), 16);
  }
  return bytes;
}

function signLowSRaw(curve, privateKey, message) {
  const { ec } = getCurve(curve);
  const msgHash = sha256(message);
  const sig = ec.sign(msgHash, privateKey);
  let [r, s] = [sig.r, sig.s];
  [r, s] = lowS(curve, r, s);
  const raw = new Uint8Array(64);
  raw.set(bigintToBytes(r, 32), 0);
  raw.set(bigintToBytes(s, 32), 32);
  return raw;
}

function derLength(length) {
  if (length < 0x80) {
    return Buffer.from([length]);
  }
  if (length <= 0xff) {
    return Buffer.from([0x81, length]);
  }
  return Buffer.from([0x82, (length >>> 8) & 0xff, length & 0xff]);
}

function derWrap(tag, content) {
  const body = Buffer.from(content);
  return Buffer.concat([Buffer.from([tag]), derLength(body.length), body]);
}

function derSequence(parts) {
  const content = Buffer.concat(parts.map((part) => Buffer.from(part)));
  return derWrap(0x30, content);
}

function derIntegerSmall(n) {
  if (n === 0) return Buffer.from([0x02, 0x01, 0x00]);
  if (n === 1) return Buffer.from([0x02, 0x01, 0x01]);
  throw new Error('unsupported integer');
}

function derOctetString(data) {
  return derWrap(0x04, data);
}

function derBitString(data) {
  return derWrap(0x03, Buffer.concat([Buffer.from([0x00]), Buffer.from(data)]));
}

function buildPkcs8Der(curve, privateKeyBytes, uncompressedPubkey) {
  const { algId } = getCurve(curve);
  const ecPrivateKey = derSequence([
    derIntegerSmall(1),
    derOctetString(privateKeyBytes),
    derWrap(0xa1, derBitString(uncompressedPubkey)),
  ]);
  return derSequence([derIntegerSmall(0), algId, derOctetString(ecPrivateKey)]);
}

function buildSpkiDer(curve, uncompressedPubkey) {
  const { algId } = getCurve(curve);
  return derSequence([algId, derBitString(uncompressedPubkey)]);
}

function toPem(label, der) {
  const b64 = Buffer.from(der).toString('base64');
  const lines = b64.match(/.{1,64}/g) ?? [];
  return `-----BEGIN ${label}-----\n${lines.join('\n')}\n-----END ${label}-----\n`;
}

function generateRotationKey(outdir, curve, verbose = false) {
  ensureDir(outdir);
  const { ec } = getCurve(curve);

  const privateKeyBytes = ec.utils.randomPrivateKey();
  const compressed = ec.getPublicKey(privateKeyBytes, true);
  const uncompressed = ec.getPublicKey(privateKeyBytes, false);

  const privPemPath = `${outdir}/rotation_${curve}_private.pem`;
  const pubPemPath = `${outdir}/rotation_${curve}_public.pem`;

  const privatePem = toPem('PRIVATE KEY', buildPkcs8Der(curve, privateKeyBytes, uncompressed));
  const publicPem = toPem('PUBLIC KEY', buildSpkiDer(curve, uncompressed));
  writeFileSync(privPemPath, privatePem);
  writeFileSync(pubPemPath, publicPem);

  const didKey = didKeyForPub(curve, compressed);

  if (verbose) {
    console.log(`[verbose] Generated ${curve.toUpperCase()} keypair`);
    console.log(`[verbose] Compressed pubkey: ${Buffer.from(compressed).toString('hex')}`);
  }

  return {
    curve,
    privPemPath,
    pubPemPath,
    didKey,
    privateKeyBytes,
    compressed,
  };
}

function buildUnsignedOp(rotationDidKeys, akaList, pdsEndpoint) {
  const verificationMethods = {};
  const services = {};
  if (pdsEndpoint) {
    services.atproto_pds = {
      type: 'AtprotoPersonalDataServer',
      endpoint: pdsEndpoint,
    };
  }
  return {
    type: 'plc_operation',
    rotationKeys: rotationDidKeys,
    verificationMethods: verificationMethods,
    alsoKnownAs: akaList,
    services: services,
    prev: null,
  };
}

function derivePlcDid(signedOp) {
  const cbor = dagCborEncode(signedOp);
  const digest = sha256(cbor);
  const suffix = base32LowerNopad(digest).slice(0, 24);
  return 'did:plc:' + suffix;
}

function collect(value, previous) {
  previous.push(value);
  return previous;
}

async function main() {
  const program = new Command();
  program
    .description('Generate & register a DID:PLC genesis operation.')
    .option('--out <dir>', 'Output directory for keys (default: plc_keys)', 'plc_keys')
    .option('--curve <curve>', 'Rotation key curve', 'k256')
    .option('--aka <uri>', 'alsoKnownAs entry (e.g., at://alice.example). May repeat.', collect, [])
    .option('--pds <url>', 'PDS endpoint URL (e.g., https://pds.example.com)')
    .option('--dry-run', 'Build & print but do not POST to PLC')
    .option('-v, --verbose', 'Show verbose output')
    .parse();

  const args = program.opts();
  if (!['k256', 'p256'].includes(args.curve)) {
    console.error(`error: option '--curve' must be 'k256' or 'p256'`);
    process.exit(1);
  }
  const kb = generateRotationKey(args.out, args.curve, args.verbose);
  const unsigned = buildUnsignedOp([kb.didKey], args.aka, args.pds ?? null);

  if (args.verbose) {
    console.log('[verbose] Unsigned operation:');
    console.log(JSON.stringify(unsigned, null, 2));
  }

  const unsignedCbor = dagCborEncode(unsigned);
  if (args.verbose) {
    console.log(`[verbose] Encoded CBOR size: ${unsignedCbor.length} bytes`);
  }

  const rawSig = signLowSRaw(args.curve, kb.privateKeyBytes, unsignedCbor);
  const sigB64u = b64urlNopad(rawSig);

  if (args.verbose) {
    console.log(`[verbose] Signature (base64url): ${sigB64u}`);
  }

  const signed = { ...unsigned, sig: sigB64u };
  const did = derivePlcDid(signed);

  if (args.verbose) {
    const signedCbor = dagCborEncode(signed);
    const digest = sha256(signedCbor);
    console.log(`[verbose] SHA256 of signed op: ${Buffer.from(digest).toString('hex')}`);
  }

  writeFileSync(`${args.out}/genesis_unsigned.dag-cbor`, Buffer.from(unsignedCbor));
  writeFileSync(`${args.out}/genesis_signed.json`, `${JSON.stringify(signed)}\n`);
  writeFileSync(`${args.out}/did.txt`, `${did}\n`);
  writeFileSync(`${args.out}/rotation_did_key.txt`, `${kb.didKey}\n`);

  if (args.verbose) {
    console.log(`[verbose] Wrote genesis_unsigned.dag-cbor (${unsignedCbor.length} bytes)`);
    console.log('[verbose] Wrote genesis_signed.json');
    console.log('[verbose] Wrote did.txt');
    console.log('[verbose] Wrote rotation_did_key.txt');
  }

  console.log(`Rotation key (did:key): ${kb.didKey}`);
  console.log(`DID (derived):          ${did}`);
  console.log(`Wrote keys & artifacts to: ${resolve(args.out)}`);

  if (args.dryRun) {
    console.log('Dry run selected; not POSTing to PLC.');
    return;
  }

  const url = `https://plc.directory/${did}`;
  if (args.verbose) {
    console.log(`[verbose] POSTing to ${url}`);
    console.log('[verbose] Request body:');
    console.log(JSON.stringify(signed, null, 2));
  }

  try {
    const resp = await fetch(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(signed),
      signal: AbortSignal.timeout(10000),
    });
    const text = await resp.text();
    console.log(`POST ${url} -> ${resp.status}`);
    console.log(text.slice(0, 5000));
    if (resp.ok) {
      console.log('Registration appears successful.');
    } else {
      console.log('Registration failed; see response above.');
    }
  } catch (e) {
    process.stderr.write(`Error POSTing to PLC: ${e}\n`);
    process.exit(2);
  }
}

await main();

File: plc_register.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Minimal DID:PLC genesis op generator + registrar.

Requirements (Python 3.9+):
  pip install cryptography dag-cbor base58 requests

Example:
  # Minimal (k256, no AKA/services):
  python plc_genesis_register.py

  # With handle and PDS (repeatable; does not upload keys):
  python plc_genesis_register.py --aka at://alice.example --pds https://pds.example.com

  # Use p256 instead of k256:
  python plc_genesis_register.py --curve p256 --aka at://bob.example --pds https://pds.example.com
"""

import argparse
import base64
import dataclasses
import hashlib
import json
import pathlib
import sys
import typing as t

import base58
import dag_cbor
import requests
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import ec, utils

# --- Constants from atproto crypto spec (multicodec varints for compressed pubkeys) ---
# p256-pub  code 0x1200  -> varint bytes [0x80, 0x24]
# secp256k1-pub code 0xE7 -> varint bytes [0xE7, 0x01]
# Ref: https://atproto.com/specs/cryptography
MC_P256_PUB = bytes([0x80, 0x24])
MC_K256_PUB = bytes([0xE7, 0x01])

# Curve orders (for low-S canonicalization)
# Ref: standards / well-known constants
N_K256 = int("FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141", 16)
N_P256 = int("FFFFFFFF00000000FFFFFFFFFFFFFFFFBCE6FAADA7179E84F3B9CAC2FC632551", 16)


def ensure_dir(path: pathlib.Path) -> None:
    path.mkdir(parents=True, exist_ok=True)


def b64url_nopad(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).rstrip(b"=").decode("ascii")


def base32_lower_nopad(data: bytes) -> str:
    # Python emits uppercase with padding; PLC uses lowercase, no padding
    return base64.b32encode(data).decode("ascii").lower().rstrip("=")


def compress_pubkey_bytes(pubkey) -> bytes:
    # cryptography supports X9.62 compressed point for p256 and k256
    return pubkey.public_bytes(
        encoding=serialization.Encoding.X962,
        format=serialization.PublicFormat.CompressedPoint,
    )


def did_key_for_pub(curve: str, pubkey_bytes_compressed: bytes) -> str:
    if curve == "k256":
        pref = MC_K256_PUB
    elif curve == "p256":
        pref = MC_P256_PUB
    else:
        raise ValueError("curve must be 'k256' or 'p256'")
    mb = b"z" + base58.b58encode(pref + pubkey_bytes_compressed)
    return "did:key:" + mb.decode("ascii")


def low_s_r_s(curve: str, r: int, s: int) -> t.Tuple[int, int]:
    if curve == "k256":
        n = N_K256
    else:
        n = N_P256
    # Enforce low-S (s <= n/2)
    if s > n // 2:
        s = n - s
    return r, s


def sign_low_s_raw(curve: str, private_key, message: bytes) -> bytes:
    # cryptography signs ECDSA over the given hash algorithm
    der_sig = private_key.sign(message, ec.ECDSA(hashes.SHA256()))
    r, s = utils.decode_dss_signature(der_sig)
    r, s = low_s_r_s(curve, r, s)
    # raw 64-byte r||s big-endian
    rb = r.to_bytes(32, "big")
    sb = s.to_bytes(32, "big")
    return rb + sb


@dataclasses.dataclass
class KeyBundle:
    curve: str  # 'k256' or 'p256'
    priv_pem_path: pathlib.Path
    pub_pem_path: pathlib.Path
    did_key: str
    signing_key_bytes: bytes  # compressed point (33 bytes)


def generate_rotation_key(outdir: pathlib.Path, curve: str, verbose: bool = False) -> KeyBundle:
    ensure_dir(outdir)
    if curve == "k256":
        priv = ec.generate_private_key(ec.SECP256K1(), backend=default_backend())
    elif curve == "p256":
        priv = ec.generate_private_key(ec.SECP256R1(), backend=default_backend())
    else:
        raise ValueError("curve must be 'k256' or 'p256'")

    pub = priv.public_key()

    # Save keys (unencrypted PEM for simplicity; protect these in real deployments)
    priv_pem = priv.private_bytes(
        encoding=serialization.Encoding.PEM,
        format=serialization.PrivateFormat.PKCS8,
        encryption_algorithm=serialization.NoEncryption(),
    )
    pub_pem = pub.public_bytes(
        encoding=serialization.Encoding.PEM,
        format=serialization.PublicFormat.SubjectPublicKeyInfo,
    )

    priv_path = outdir / f"rotation_{curve}_private.pem"
    pub_path = outdir / f"rotation_{curve}_public.pem"
    priv_path.write_bytes(priv_pem)
    pub_path.write_bytes(pub_pem)

    # did:key for rotation key is allowed;
    # PLC only requires rotationKeys to be did:key of k256 or p256
    compressed = compress_pubkey_bytes(pub)
    didk = did_key_for_pub(curve, compressed)

    if verbose:
        print(f"[verbose] Generated {curve.upper()} keypair")
        print(f"[verbose] Compressed pubkey: {compressed.hex()}")

    return KeyBundle(
        curve=curve,
        priv_pem_path=priv_path,
        pub_pem_path=pub_path,
        did_key=didk,
        signing_key_bytes=compressed,
    )


def build_unsigned_op(
    rotation_did_keys: t.List[str], aka_list: t.List[str], pds_endpoint: t.Optional[str]
) -> dict:
    # Minimal regular op shape; empty maps/lists are fine
    verification_methods = {}
    services = {}
    if pds_endpoint:
        services["atproto_pds"] = {
            "type": "AtprotoPersonalDataServer",
            "endpoint": pds_endpoint,
        }
    return {
        "type": "plc_operation",
        "rotationKeys": rotation_did_keys,
        "verificationMethods": verification_methods,
        "alsoKnownAs": aka_list,
        "services": services,
        "prev": None,  # must be present and null for genesis
        # 'sig' inserted after signing
    }


def derive_plc_did(signed_op: dict) -> str:
    cbor = dag_cbor.encode(signed_op)
    digest = hashlib.sha256(cbor).digest()
    suffix = base32_lower_nopad(digest)[:24]
    return "did:plc:" + suffix


def main():
    ap = argparse.ArgumentParser(description="Generate & register a DID:PLC genesis operation.")
    ap.add_argument(
        "--out", default="plc_keys", help="Output directory for keys (default: plc_keys)"
    )
    ap.add_argument("--curve", choices=["k256", "p256"], default="k256", help="Rotation key curve")
    ap.add_argument(
        "--aka",
        action="append",
        default=[],
        help="alsoKnownAs entry (e.g., at://alice.example). May repeat.",
    )
    ap.add_argument("--pds", default=None, help="PDS endpoint URL (e.g., https://pds.example.com)")
    ap.add_argument("--dry-run", action="store_true", help="Build & print but do not POST to PLC")
    ap.add_argument("-v", "--verbose", action="store_true", help="Show verbose output")
    args = ap.parse_args()

    outdir = pathlib.Path(args.out)
    kb = generate_rotation_key(outdir, args.curve, args.verbose)

    unsigned = build_unsigned_op([kb.did_key], args.aka, args.pds)

    if args.verbose:
        print("[verbose] Unsigned operation:")
        print(json.dumps(unsigned, indent=2))

    # Sign DAG-CBOR bytes of unsigned op (without 'sig')
    unsigned_cbor = dag_cbor.encode(unsigned)
    if args.verbose:
        print(f"[verbose] Encoded CBOR size: {len(unsigned_cbor)} bytes")

    # cryptography ECDSA expects the message; spec requires ECDSA-SHA256;
    # library will hash internally. To sign the SHA256 digest explicitly,
    # replace message with hashlib.sha256(unsigned_cbor).digest()
    # and switch to ec.Prehashed(hashes.SHA256()).
    # Low-S enforcement + raw r||s per atproto crypto guidance.
    priv = serialization.load_pem_private_key(kb.priv_pem_path.read_bytes(), password=None)
    raw_sig = sign_low_s_raw(args.curve, priv, unsigned_cbor)
    sig_b64u = b64url_nopad(raw_sig)

    if args.verbose:
        print(f"[verbose] Signature (base64url): {sig_b64u}")

    signed = dict(unsigned)
    signed["sig"] = sig_b64u

    did = derive_plc_did(signed)

    if args.verbose:
        signed_cbor = dag_cbor.encode(signed)
        digest = hashlib.sha256(signed_cbor).digest()
        print(f"[verbose] SHA256 of signed op: {digest.hex()}")

    # Save artifacts
    (outdir / "genesis_unsigned.dag-cbor").write_bytes(unsigned_cbor)
    (outdir / "genesis_signed.json").write_text(
        json.dumps(signed, separators=(",", ":"), ensure_ascii=False) + "\n"
    )
    (outdir / "did.txt").write_text(did + "\n")
    (outdir / "rotation_did_key.txt").write_text(kb.did_key + "\n")

    if args.verbose:
        print(f"[verbose] Wrote genesis_unsigned.dag-cbor ({len(unsigned_cbor)} bytes)")
        print("[verbose] Wrote genesis_signed.json")
        print("[verbose] Wrote did.txt")
        print("[verbose] Wrote rotation_did_key.txt")

    print(f"Rotation key (did:key): {kb.did_key}")
    print(f"DID (derived):          {did}")
    print(f"Wrote keys & artifacts to: {outdir.resolve()}")

    if args.dry_run:
        print("Dry run selected; not POSTing to PLC.")
        return

    # Submit to PLC directory (HTTP POST JSON to .../{did})
    url = f"https://plc.directory/{did}"
    if args.verbose:
        print(f"[verbose] POSTing to {url}")
        print("[verbose] Request body:")
        print(json.dumps(signed, indent=2))

    try:
        resp = requests.post(url, json=signed, timeout=10)
        print(f"POST {url} -> {resp.status_code}")
        print(resp.text[:5000])
        if 200 <= resp.status_code < 300:
            print("Registration appears successful.")
        else:
            print("Registration failed; see response above.")
    except Exception as e:
        print(f"Error POSTing to PLC: {e}", file=sys.stderr)
        sys.exit(2)


if __name__ == "__main__":
    main()
